src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs

```diff
@@ -84,13 +70,48 @@ public void OnSolutionRemoved()
                 UnregisterPrimarySolution(solutionId, synchronousShutdown: false);
             }
 
+            private void RegisterPrimarySolution()
+            {
+                this.AssertIsForeground();
+                _currentSolutionId = _workspace.CurrentSolution.Id;
+                var solutionId = _currentSolutionId;
+
+                var storageLocation = _workspace.DeferredState?.ProjectTracker.GetWorkingFolderPath(_workspace.CurrentSolution);
+
+                try
+                {
+                    // this will only return false if OOP doesn't exist. otherwise,
+                    // it will throw and handled as usual
+                    _session.TryInvokeAsync(
+                        nameof(IRemoteHostService.RegisterPrimarySolutionId),
+                        new object[] { solutionId, storageLocation }, CancellationToken.None).Wait(CancellationToken.None);
+                }
+                catch (RemoteHostClientExtensions.UnexpectedRemoteHostException)
+                {
+                    // ignore unexpected remote host exception. it is allowed here since it is part of OOP engine.
+                    // no one outside of engine should ever catch this exception or care about it.
```

Comment:
You say no one outside of engine shoudl ever catch this exception... but it looks like we're outside the engine, and catching it.  What constitutes "engine" in this description?

Reply:
see the filename. it is still inside of OOP engine.

Reply:
I concur with @CyrusNajmabadi here: I can't actually tell what is considered internal and public and anything else here. If the exception type was some private class of something else I'd feel better that it's "internal" but right now somebody can take a dependency on this.

Reply:
due to layering that is not possible. I can make it internal, but I made it nested type of the extensions for that.

I still not getting that reasoning. unless all OOP types are moved to its own dll and make only those APIs that are supposed to be used by others are marked public, I am not sure how I understand your comment.

for example, ServiceHubRemoteHostClient is something nobody in Roslyn should ever create themselves. it is one of implementation of IRemoteHostClient. but since people can see everything in Roslyn, they can do it. and break whole system by creating another OOP process which lacks all internal states.

same goes for all other services, such as VisualStudioMetadataProvider and all which will break metadata identity, CompilationTracker which will break symbol's identity.

as Roslyn developer, it is their responsibility what is boundary of those internal types. it is cost of all Roslyn being internal visible to all its internal types and dlls.

Reply:
> see the filename. it is still inside of OOP engine

What is the 'engine' in this design?  It's really not clear to me.  

> same goes for all other services, such as VisualStudioMetadataProvider

Our other services don't throw specific exceptions that are only supposed to be handled by some codepaths, but not by others. It's really hard to tell if this is right or not because there's no way to really know if you've gotten the entire list of callers that need protection.  Also, because the layers of the remote system aren't clear, it's not easy (or possible) for someone else to come in and make a determination if other locations are a problem.

Reply:
we are talking about internal types boundary. we are switching subject.

we know it is not clear. and that's fine since that was clearly communicated to the manager when this is added. consensus at the time was, as long as it make VS not hard crash, but soft crash most of time, and give users time to save their work, it is fine. it proved that did its work. whether it is going to work for 100%, no idea. and that's work for NFW change on Roslyn or moving back to crash behavior once most of issues we know got fixed.

until then, this is what we have. and this PR is fixing a few specific cases we know this workaround doesn't work. 

and the exception is exposed since it is inside of same OOP engine. it is written in comment again and again to indicate that. 

you need to read code to see layering. are there any complex features that have simple layering? see Options, Editor work for VS4Mac, editorconfig and etc.